scipy/stats/_sobol.pyx

```diff
@@ -195,14 +191,18 @@ cdef int ibits(const int x, const int pos, const int length) nogil:
 
 @cython.boundscheck(False)
 @cython.wraparound(False)
-cpdef void initialize_v(cnp.int_t[:, :] v, const int dim):
-    cdef int d, i, j, k, m, p, newv, pow2
+cpdef void _initialize_v(
+    cnp.uint64_t[:, ::1] v, const int dim, const int maxbit
```

Comment:
Again, I'll trust this. But if you don't see a performance difference, I don't see why to change it, especially here.

Reply:
On my platform I don't see any change of performance, but I have a M1 which has a rather different memory architecture. Maybe someone with a "classical" architecture would see a benefit. From Cython doc, they recommend to do this and it should bring around 10% performance improvements.

@tirthasheshpatel proposed to check on his side.

Reply:
I did some experiments. Contiguous memoryviews don't change anything other than how indexing is done in C. So, I wrote a simple Cython extension with a non-contiguous memoryview:

```cython
import numpy as np
cimport numpy as np

np.import_array()

cdef np.uint64_t [:] my_arr
my_arr = np.zeros(5)

print(my_arr[1])
```

The generated C file uses the strides to index elements of the array:

```c
__pyx_v_12memory_views_my_arr.data + __pyx_t_5 * __pyx_v_12memory_views_my_arr.strides[0]
```

While, when I change the line `cdef np.uint64_t [:] my_arr` to `cdef np.uint64_t [::1] my_arr`, the only change is:

```c
__pyx_v_12memory_views_my_arr.data) + __pyx_t_5
```

i.e. strides are no longer used when indexing. If you have a large for loop, this could speed up things by 2-5% but doesn't give a great benefit.

The same is applicable to multidimensional array (only different being the speedup is even smaller)

<details>

<summary>Diff when memoryview is changed to contiguous memoryview in my little example</summary>

Cython version 0.29.28
NumPy version 1.22.2

```diff
diff --git a/cython_stuff/memory_views.c b/cython_stuff/memory_views.c
index df9b7d3..a435437 100644
--- a/cython_stuff/memory_views.c
+++ b/cython_stuff/memory_views.c
@@ -2024,7 +2024,7 @@ static int __Pyx_ValidateAndInit_memviewslice(
                 PyObject *original_obj);
 
 /* ObjectToMemviewSlice.proto */
-static CYTHON_INLINE __Pyx_memviewslice __Pyx_PyObject_to_MemoryviewSlice_ds_nn___pyx_t_5numpy_uint64_t(PyObject *, int writable_flag);
+static CYTHON_INLINE __Pyx_memviewslice __Pyx_PyObject_to_MemoryviewSlice_dc_nn___pyx_t_5numpy_uint64_t(PyObject *, int writable_flag);
 
 /* GCCDiagnostics.proto */
 #if defined(__GNUC__) && (__GNUC__ > 4 || (__GNUC__ == 4 && __GNUC_MINOR__ >= 6))
@@ -17676,7 +17676,7 @@ static CYTHON_SMALL_CODE int __Pyx_InitCachedConstants(void) {
 
   /* "memory_views.pyx":7
  * 
- * cdef np.uint64_t [:] my_arr
+ * cdef np.uint64_t [::1] my_arr
  * my_arr = np.zeros(5)             # <<<<<<<<<<<<<<
  * 
  * print(my_arr[1])
@@ -18166,13 +18166,13 @@ if (!__Pyx_RefNanny) {
  * 
  * np.import_array()             # <<<<<<<<<<<<<<
  * 
- * cdef np.uint64_t [:] my_arr
+ * cdef np.uint64_t [::1] my_arr
  */
   __pyx_t_2 = __pyx_f_5numpy_import_array(); if (unlikely(__pyx_t_2 == ((int)-1))) __PYX_ERR(0, 4, __pyx_L1_error)
 
   /* "memory_views.pyx":7
  * 
- * cdef np.uint64_t [:] my_arr
+ * cdef np.uint64_t [::1] my_arr
  * my_arr = np.zeros(5)             # <<<<<<<<<<<<<<
  * 
  * print(my_arr[1])
@@ -18185,7 +18185,7 @@ if (!__Pyx_RefNanny) {
   __pyx_t_1 = __Pyx_PyObject_Call(__pyx_t_3, __pyx_tuple__21, NULL); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 7, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_1);
   __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
-  __pyx_t_4 = __Pyx_PyObject_to_MemoryviewSlice_ds_nn___pyx_t_5numpy_uint64_t(__pyx_t_1, PyBUF_WRITABLE); if (unlikely(!__pyx_t_4.memview)) __PYX_ERR(0, 7, __pyx_L1_error)
+  __pyx_t_4 = __Pyx_PyObject_to_MemoryviewSlice_dc_nn___pyx_t_5numpy_uint64_t(__pyx_t_1, PyBUF_WRITABLE); if (unlikely(!__pyx_t_4.memview)) __PYX_ERR(0, 7, __pyx_L1_error)
   __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
   __PYX_XDEC_MEMVIEW(&__pyx_v_12memory_views_my_arr, 1);
   __pyx_v_12memory_views_my_arr = __pyx_t_4;
@@ -18208,7 +18208,7 @@ if (!__Pyx_RefNanny) {
     __Pyx_RaiseBufferIndexError(__pyx_t_2);
     __PYX_ERR(0, 9, __pyx_L1_error)
   }
-  __pyx_t_1 = __Pyx_PyInt_From_npy_uint64((*((__pyx_t_5numpy_uint64_t *) ( /* dim=0 */ (__pyx_v_12memory_views_my_arr.data + __pyx_t_5 * __pyx_v_12memory_views_my_arr.strides[0]) )))); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 9, __pyx_L1_error)
+  __pyx_t_1 = __Pyx_PyInt_From_npy_uint64((*((__pyx_t_5numpy_uint64_t *) ( /* dim=0 */ ((char *) (((__pyx_t_5numpy_uint64_t *) __pyx_v_12memory_views_my_arr.data) + __pyx_t_5)) )))); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 9, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_1);
   __pyx_t_3 = __Pyx_PyObject_CallOneArg(__pyx_builtin_print, __pyx_t_1); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 9, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_3);
@@ -21412,17 +21412,17 @@ no_fail:
 }
 
 /* ObjectToMemviewSlice */
-  static CYTHON_INLINE __Pyx_memviewslice __Pyx_PyObject_to_MemoryviewSlice_ds_nn___pyx_t_5numpy_uint64_t(PyObject *obj, int writable_flag) {
+  static CYTHON_INLINE __Pyx_memviewslice __Pyx_PyObject_to_MemoryviewSlice_dc_nn___pyx_t_5numpy_uint64_t(PyObject *obj, int writable_flag) {
     __Pyx_memviewslice result = { 0, 0, { 0 }, { 0 }, { 0 } };
     __Pyx_BufFmt_StackElem stack[1];
-    int axes_specs[] = { (__Pyx_MEMVIEW_DIRECT | __Pyx_MEMVIEW_STRIDED) };
+    int axes_specs[] = { (__Pyx_MEMVIEW_DIRECT | __Pyx_MEMVIEW_CONTIG) };
     int retcode;
     if (obj == Py_None) {
         result.memview = (struct __pyx_memoryview_obj *) Py_None;
         return result;
     }
-    retcode = __Pyx_ValidateAndInit_memviewslice(axes_specs, 0,
-                                                 PyBUF_RECORDS_RO | writable_flag, 1,
+    retcode = __Pyx_ValidateAndInit_memviewslice(axes_specs, __Pyx_IS_C_CONTIG,
+                                                 (PyBUF_C_CONTIGUOUS | PyBUF_FORMAT) | writable_flag, 1,
                                                  &__Pyx_TypeInfo_nn___pyx_t_5numpy_uint64_t, stack,
                                                  &result, obj);
     if (unlikely(retcode == -1))
diff --git a/cython_stuff/memory_views.pyx b/cython_stuff/memory_views.pyx
index cd14a99..8ac389d 100644
--- a/cython_stuff/memory_views.pyx
+++ b/cython_stuff/memory_views.pyx
@@ -3,7 +3,7 @@ cimport numpy as np
 
 np.import_array()
 
-cdef np.uint64_t [:] my_arr
+cdef np.uint64_t [::1] my_arr
 my_arr = np.zeros(5)
 
 print(my_arr[1])
```

</details>

Reply:
Thanks @tirthasheshpatel for the analysis. Well then if there is no real drawback, 2-5% in some cases is good to have.